proof-generator/src/lib.rs

```diff
@@ -0,0 +1,174 @@
+    NoBlockHash,
+    /// ETH signature failed
+    #[error("eth signature failed")]
+    ETHSigFailed,
+    /// GPG signature failed
+    #[error("{0}")]
+    GPGSigFailed(String),
+}
+
+/// Sign a message with a GPG private key using the GPG CLI
+fn gpg_sign(key: &str, proof: &str) -> anyhow::Result<String> {
+    let mut gpg = Command::new("gpg")
+        .arg("--clear-sign")
+        .arg("-u")
+        .arg(key)
+        .stdin(Stdio::piped())
+        .stderr(Stdio::piped())
+        .stdout(Stdio::piped())
+        .spawn()?;
+
+    gpg.stdin.as_mut().unwrap().write_all(proof.as_bytes())?;
+
+    let output = gpg.wait_with_output()?;
+    if output.status.success() {
+        let raw_output = String::from_utf8(output.stdout)?;
+        Ok(raw_output)
+    } else {
+        Err(anyhow!(Error::GPGSigFailed(String::from_utf8(
+            output.stderr
+        )?)))
+    }
+}
+
+/// Sign a message with a ETH private key using either a keystore file or a Ledger HW
+pub async fn eth_sign<S: 'static + Signer>(
+    signer: &S,
+    proof: &str,
+) -> Result<Signature, <S as Signer>::Error> {
+    signer.sign_message(proof).await
+}
+
+/// Create the proof to be signed
+fn create_proof<T: Display, K: Debug>(
+    ownership: &T,
+    evidence: &K,
+    block_hash: &H256,
+) -> Result<String, Error> {
+    Ok(format!(
+        "As the owner of GPG key {}, my Ethereum address is {:?} as of {:?}",
+        &ownership, &evidence, &block_hash
+    ))
+}
+
+/// Sign the created proof with GPG and ETH keypairs
+async fn sign_proof<T: 'static + Signer>(
+    gpg_key: &str,
+    signer: &T,
+    block_hash: &H256,
+) -> anyhow::Result<Proof> {
+    let msg = create_proof(&gpg_key, &signer.address(), &block_hash)?;
+
+    log::info!("Signing message with ETH keypair..");
+    let eth_sig = eth_sign(signer, &msg)
+        .await
+        .map_err(|_| anyhow!(Error::ETHSigFailed))?;
+    log::debug!("ETH Signature: {:?}.", eth_sig);
+
+    log::info!("Signing message with GPG keypair..");
+    let gpg_sig = gpg_sign(&gpg_key, &msg)?;
+    log::debug!("GPG Signature: {:?}.", gpg_sig);
+
+    Ok(Proof { gpg_sig, eth_sig })
+}
+
+/// The main lib function that runs the functionality of the program
+/// - Obtains a block hash from a block from 1 day ago.
+/// - Gets either a keystore file or in its absence a Ledger HW as signer to sign a message.
+/// - Creates a message that will be signed by the defined signer.
+/// - Write both proofs to a JSON file.
+pub async fn run(opts: Options) -> anyhow::Result<()> {
+    let provider =
+        Provider::<Http>::try_from(opts.rpc_url).expect("could not instantiate HTTP Provider");
+    let latest_block_number = provider.get_block_number().await?;
+    // 5760 blocks earlier is aprox. 1 day ago, this due to avoid referencing blocks that are affected by reorgs of the chain.
+    let block_number = latest_block_number - 5760u64;
+    if let Some(block) = provider.get_block(block_number).await? {
```

Comment:
You can flatten these nested conditionals by using `ok_or` https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or